feat(rewardService): add reward call retries

rickstaa opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

This pull request refactors the RewardService to create a retry worker with exponential backoff for retrying Reward calls. Retries are limited to a specified number and only attempted for the intended round, ensuring efficiency and avoiding redundant calls.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- Update the retry logic to retry failed rewards.
- Updates tests to check the new code.
- Add a new `RewardRetryTimes` parameter for people to set the number of retries.

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

I wrote tests. I still have to test them on a real orchestrator.

**Does this pull request close any open issues?**
<!-- Fixes # -->

No.

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./CONTRIBUTING.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

rickstaa committed:
feat(rewardService): add reward call retries

This commit adds logic to retry reward calls if they fail. The retry mechanism
can be enabled by setting the `rewardRetryInterval` to a value greater than 0.
If `rewardRetryInterval` is set to 0, retries are disabled, and the reward call
is attempted only once.

rickstaa commented:
Just me doing some experimentation.

ad-astra-video commented:
Couple quick suggestions:
- I would change `rewardRetryInterval` to something that limits the amount of retries (maybe `rewardRetryTimes`).  I would not want an unbounded amount of retries possible.  L2 gas fees are cheap but could drain the ETH from the wallet slowly.
- Use a single interval like 13-26 seconds to retry reward call in a new ethereum L1 block to get past gas fees on the current block
- I can see a case to have both flags for retry interval and number of retries to do to give more control how often and how many times retries are tried.

rickstaa commented:
Some concerns given by @thomshutt:
> Looks really good, my only concern would be that the retries are unbounded. I think I'd actually prefer to just have it retry X (5?) times by default with a reasonable interval rather than exposing it as yet another parameter, since I don't think people gain much from tweaking the time between retries

rickstaa committed:
refactor: limit reward call retries and improve retry logic

Limits the amount of retries done and ensure retries are only attempted
for the intended round, avoiding redundant reward calls. Updated tests
to validate the new retry.

rickstaa committed:
chore: update pending changelog

leszko commented on eth/rewardservice.go:
```diff
@@ -85,6 +94,25 @@ func (s *RewardService) IsWorking() bool {
 	return s.working
 }
 
+func (s *RewardService) startNewWorker(ctx context.Context, round int64) {
```
You could consider having one loop and one goroutine instead of starting a new goroutine each time you receive a reward request. I think this would be slightly more go-style and readable as well. Anyway, also not pretty, so up to you.

```golang
s.rewardCh = make(chan struct{}, 1)
var backoffCancel context.CancelFunc

go func() {
	for {
		select {
		case <-s.rewardCh:
			// Cancel any ongoing backoff/retry loop
			if backoffCancel != nil {
				backoffCancel()
			}
			var ctx context.Context
			ctx, backoffCancel = context.WithCancel(context.Background())

			go func(ctx context.Context) {
				expBackoff := backoff.NewExponentialBackOff()
				expBackoff.InitialInterval = s.retryInterval
				expBackoff.MaxInterval = 1 * time.Hour
				expBackoff.MaxElapsedTime = s.maxElapsedTime
				backoffCtx := backoff.WithContext(expBackoff, ctx)

				for {
					err := s.tryReward()
					if err == nil {
						return
					}
					select {
					case <-backoff.After(backoffCtx):
						// continue retrying
					case <-ctx.Done():
						return
					}
				}
			}(ctx)
		}
	}
}()
```

leszko reviewed (approved):
Added one inline comment, other than that it looks good. 

Could please also?
1. Update the description given the reason why the reward actually fails from time to time? In other words, to explain why we even need a retry 🙃 
2. Test if it works 😄
